modin/experimental/core/execution/native/implementations/omnisci_on_native/exchange/dataframe_protocol/dataframe.py

```diff
@@ -0,0 +1,326 @@
+            OmnisciProtocolDataframe(
+                self._df.mask(col_positions=[i]),
+                allow_copy=self._allow_copy,
+            ),
+        )
+
+    def get_column_by_name(self, name: str) -> OmnisciProtocolColumn:
+        return OmnisciProtocolColumn(
+            OmnisciProtocolDataframe(
+                self._df.mask(col_labels=[name]),
+                allow_copy=self._allow_copy,
+            ),
+        )
+
+    def get_columns(self) -> Iterable[OmnisciProtocolColumn]:
+        for name in self._df.columns:
+            yield OmnisciProtocolColumn(
+                OmnisciProtocolDataframe(
+                    self._df.mask(col_labels=[name]),
+                    nan_as_null=self._nan_as_null,
+                    allow_copy=self._allow_copy,
+                ),
+            )
+
+    def select_columns(self, indices: Sequence[int]) -> "OmnisciProtocolDataframe":
+        if not isinstance(indices, collections.Sequence):
+            raise ValueError("`indices` is not a sequence")
+
+        return OmnisciProtocolDataframe(
+            self._df.mask(col_positions=list(indices)),
+            nan_as_null=self._nan_as_null,
+            allow_copy=self._allow_copy,
+        )
+
+    def select_columns_by_name(
+        self, names: Sequence[str]
+    ) -> "OmnisciProtocolDataframe":
+        if not isinstance(names, collections.Sequence):
+            raise ValueError("`names` is not a sequence")
+
+        return OmnisciProtocolDataframe(
+            self._df.mask(col_labels=list(names)),
+            nan_as_null=self._nan_as_null,
+            allow_copy=self._allow_copy,
+        )
+
+    def get_chunks(
+        self, n_chunks: Optional[int] = None
+    ) -> Iterable["OmnisciProtocolDataframe"]:
+        if n_chunks is None:
+            return self._yield_chunks(self._chunk_slices)
+
+        if n_chunks % self.num_chunks() != 0:
+            raise RuntimeError(
+                "The passed `n_chunks` has to be a multiple of `num_chunks`."
+            )
+
+        if n_chunks > self.num_rows():
+            raise RuntimeError(
+                "The passed `n_chunks` value is bigger than the amout of rows in the frame."
+            )
+
+        extra_chunks = n_chunks - self.num_chunks()
+        subdivided_slices = self._chunk_slices.copy()
+
+        # The subdividing behavior is a bit different from "subdividing each chunk",
+        # instead it subdivides the biggest chunks first, so overall chunking be as
+        # equal as possible
+        for _ in range(extra_chunks):
+            # 1. Find the biggest chunk
+            # 2. Split it in the middle
+            biggest_chunk_idx = np.argmax(np.diff(subdivided_slices))
+            new_chunk_offset = (
+                subdivided_slices[biggest_chunk_idx + 1]
+                - subdivided_slices[biggest_chunk_idx]
+            ) // 2
+            if new_chunk_offset == 0:
+                raise RuntimeError("No more chunks to subdivide.")
```

Comment:
The protocol does say nothing about this exception so I think we should remove it.

Reply:
Is it better we get an undefined behavior in this case than an error with a little hint of what went wrong? This case is more about raising an internal error, this should never happen

Reply:
replaced this with calling `ErrorMessage.catch_bugs_and_request_emails` that raises an internal error

Reply:
If this should never happen, why should we have it? Can we replace the logic of subdividing to other one?

Reply:
> If this should never happen, why should we have it?

Just safety reasons, in case there's some mistake in the algorithm that we haven't caught neither on review nor in testing.

Reply:
What about the second question? I still feel that throwing an exception isn't user-friendly if all the requirements to the method are complied.

Reply:
The exception won't occur unless there's a bug in functions that the `get_chunks()` is using.

The `new_chunk_offset` is computed the following way:
```python
new_chunk_offset = (
    subdivided_slices[biggest_chunk_idx + 1]
    - subdivided_slices[biggest_chunk_idx]
) // 2
```
The only way of getting it equal to zero is to have the slices difference equal to 1. The `subdivided_slices` describes the chunk offsets (global start-stop positions), meaning that it has the following properties:
1. The sequence is monotonically increasing (`arr[i] < arr[i+1]`)
2. Its length is not more than (num_rows + 1), as there can't be more chunks than the number of rows.
3. The sum of differences equals to `num_rows` (`sum(arr[i+1] - arr[i])`). This is met, as the difference describes the chunks lengths and the sum of chunk lengths has to be equal to `num_rows`.

Considering these requirements, if the biggest `subdivided_slices` difference is 1, then all of the differences are equal to 1, meaning that there's already `num_rows` amount of chunks each of them consist of one element. An intention to create one more chunk, means that the function's requirement wasn't met (`n_chunks <= self.num_rows`).

So hitting this internal-error exception should be impossible, as there's already a [check that verifies function's requirements](https://github.com/dchigarev/modin/blob/854c91a4638a385b6ab23a6bfa89221e92d0921d/modin/experimental/core/execution/native/implementations/omnisci_on_native/exchange/dataframe_protocol/dataframe.py#L279-L282). Hitting it could be possible if there's a bug in either numpy or other protocol methods that are used by `.get_chunks()`.

Reply:
Let's leave it 😄